llvm/lib/ObjCopy/ELF/ELFObject.cpp

```diff
@@ -1234,6 +1234,13 @@ static bool compareSegmentsByOffset(const Segment *A, const Segment *B) {
     return true;
   if (A->OriginalOffset > B->OriginalOffset)
     return false;
+  // If one is PT_LOAD and the other isn't (e.g. PT_INTERP, PT_GNU_RELRO,
+  // PT_TLS), order the PT_LOAD first to ensure ParentSegment relationship will
+  // be correct.
+  bool AIsLOAD = A->Type == PT_LOAD;
+  bool BIsLOAD = B->Type == PT_LOAD;
+  if (AIsLOAD != BIsLOAD)
```

Comment:
I don't know of a concrete example of this in the wild, but should we also handle the case where both are `PT_LOAD` somehow? I feel like just picking whichever is higher alignment would be enough to properly preserve the parent/child relationship and get the alignment correct.

Reply:
I've testst the fix, yes, the fix is correct

Reply:
A program is broken if it has two `PT_LOAD` at the same offset. Most dynamic loaders will translate `PT_LOAD` to mmap commands. The map from a later `PT_LOAD` will overwrite the map from a previous one. I feel that testing this would be nice, but the value is quite low, since it is very difficult to create two `PT_LOAD` at the same offset even with a linker script... And the coverage is less relevant to this patch.

Just comparing `p_align` doesn't work when PT_TLS/PT_GNU_RELRO has an alignment of maxpagesize, as my description says:

> The same issue occurs for PT_TLS/PT_GNU_RELRO if we PT_TLS's alignment is smaller and we place the PT_LOAD after PT_TLS/PT_GNU_RELRO segments (not linker default, but possible with a PHDRS linker script command).

Reply:
Sorry for the delayed response: I'm not sure I follow why simply picking the highest alignment in a set of overlapping segments isn't sufficient? The comment you made is about PT_TLS having a lower alignment than PT_LOAD, in which case we'd want the PT_LOAD's higher alignment. But actually, the type of the PT_LOAD is somewhat irrelevant I'd have thought: you could imagine a custom segment type that requires alignment overlapping another segment type and the higher alignment would still be desirable to be used, regardless of the order of those segments.

Reply:
I think I special cased `PT_LOAD` because in practice, for all segments that may overlap with a `PT_LOAD`, treating `PT_LOAD` as the parent/ancestor will work and the parent relationship is conceptually treated as such. Giving it another thought, when a `PT_LOAD` and a `PT_TLS` have the same offset and alignment, it doesn't matter if we make `PT_TLS` the parent. The algorithm will still work as intended.

Switched to the `return A->Align > B->Align;` rule.